crates/stackable-operator/src/kvp/mod.rs

```diff
@@ -90,265 +94,102 @@ where
 
-        Ok(Self(pairs))
-    }
+    /// Returns whether the list contains a key/value pair with a specific [`Key`].
+    ///
+    /// Returns `false` if `key` cannot be parsed as a valid [`Key`].
+    // TODO: Does anyone actually use this API?
+    fn contains_str_key(&self, key: &str) -> bool;
 }
-
-impl<T> From<KeyValuePairs<T>> for BTreeMap<String, String>
-where
-    T: Value,
-{
-    fn from(value: KeyValuePairs<T>) -> Self {
-        value
-            .iter()
-            .map(|pair| (pair.key().to_string(), pair.value().to_string()))
+impl<V: Value> KeyValuePairsExt for KeyValuePairs<V> {
+    fn to_unvalidated(&self) -> BTreeMap<String, String> {
+        self.iter()
+            .map(|(key, value)| (key.to_string(), value.to_string()))
             .collect()
     }
-}
-
-impl<T> Deref for KeyValuePairs<T>
-where
-    T: Value,
-{
-    type Target = BTreeSet<KeyValuePair<T>>;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-
-impl<T> KeyValuePairs<T>
-where
-    T: Value + std::default::Default,
-{
-    /// Creates a new empty list of [`KeyValuePair`]s.
-    pub fn new() -> Self {
-        Self::default()
-    }
-
-    /// Creates a new list of [`KeyValuePair`]s from `pairs`.
-    pub fn new_with(pairs: BTreeSet<KeyValuePair<T>>) -> Self {
-        Self(pairs)
-    }
-
-    /// Extends `self` with `other`.
-    pub fn extend(&mut self, other: Self) {
-        self.0.extend(other.0);
-    }
-
-    /// Inserts a new [`KeyValuePair`] into the list of pairs.
-    ///
-    /// This function overwrites any existing key/value pair. To avoid
-    /// overwriting existing pairs, either use [`KeyValuePairs::contains`] or
-    /// [`KeyValuePairs::contains_key`] before inserting or try to insert
-    /// fallible via [`KeyValuePairs::try_insert`].
-    pub fn insert(&mut self, kvp: KeyValuePair<T>) -> &mut Self {
-        self.0.insert(kvp);
-        self
-    }
-
-    /// Tries to insert a new [`KeyValuePair`] into the list of pairs.
-    ///
-    /// If the list already had this pair present, nothing is updated, and an
-    /// error is returned.
-    pub fn try_insert(&mut self, kvp: KeyValuePair<T>) -> Result<(), KeyValuePairsError> {
-        ensure!(!self.0.contains(&kvp), PairAlreadyExistsSnafu);
-        self.insert(kvp);
-        Ok(())
-    }
-
-    /// Returns if the list contains a specific [`KeyValuePair`].
-    pub fn contains(&self, kvp: impl TryInto<KeyValuePair<T>>) -> bool {
-        let Ok(kvp) = kvp.try_into() else {
-            return false;
-        };
-        self.0.contains(&kvp)
-    }
 
-    /// Returns if the list contains a key/value pair with a specific [`Key`].
-    pub fn contains_key(&self, key: impl TryInto<Key>) -> bool {
-        let Ok(key) = key.try_into() else {
+    fn contains_str_key(&self, key: &str) -> bool {
+        // We could avoid this clone by providing an UnvalidatedKeyRef and ensure that Key: Borrow<UnvalidatedKeyRef>
```

Comment:
question: Out of curiosity, where is the clone?

Reply:
`.parse::<Key>()` calls `Key`'s `FromStr`, which in turn calls `KeyPrefix` and `KeyName`s' `FromStr`s, which in turn call `ToString` (which is morally equivalent to `Clone` for `&str`, and worse for types that require actual formatting logic).